### PR TITLE
Azure database migration service template

### PR DESCRIPTION
**Pull request by ash1625**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 101-azure-database-migration-service/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,594 @@
+
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: format the json, e.g. SHIFT+ALT+F in VS Code

File: 101-azure-database-migration-service/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+                "description": "Administrator Password for the Target Azure SQL DB Server."
+            }
+        },
+	    "_artifactsLocation": 
+		{
+		    "type": "string",
+		    "metadata": {
+			    "description": "The base URI where artifacts required by this template are located including a trailing '/'"
+		    },
+		    "defaultValue": "https://datamigrationwebpi.blob.core.windows.net/templatefiles/"
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: Artifacts need to come from this repo

File: 101-azure-database-migration-service/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    },
+    "variables":
+    {
+	    "DMSServiceName": "[concat('DMS', uniqueString(resourceGroup().id))]",
+		"sourceServerName": "[take(concat('Source', uniqueString(resourceGroup().id)), 15)]",
+		"targetServerName": "[concat('targetservername', uniqueString(resourceGroup().id))]",
+        "scriptLocation": "AddDatabaseToSqlServer.ps1 ",
+		"bakFileLocation": "AdventureWorks2016.bak",
+		"bacpacFileLocation": "AdventureWorks2016.bacpac",
+        "scriptFiles": "[createarray(concat('https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-azure-database-migration-service/', variables('scriptLocation')), concat('https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-azure-database-migration-service/', variables('bakFileLocation')))]",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: use _artifactsLocation

File: 101-azure-database-migration-service/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+                        {
+                            "enabled": true
+                        }
+                    },
+                    "keySource": "Microsoft.Storage"
+                }
+            }
+        },
+        {
+            "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: update apiVersion

File: 101-azure-database-migration-service/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+            {
+                "sourcePlatform": "SQL",
+                "targetPlatform": "SQLDB"
+            },
+            "dependsOn": [
+                "[resourceId('Microsoft.DataMigration/services', variables('DMSServiceName'))]"
+            ]
+        },
+        {
+            "type": "Microsoft.Network/virtualNetworks/subnets",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: move this to be a child resource

File: 101-azure-database-migration-service/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+                            "name": "Import",
+                            "type": "extensions",
+                            "apiVersion": "2014-04-01-preview",
+                            "dependsOn":
+                            [
+                                "[concat('Microsoft.Sql/servers/', concat(variables('targetServerName')), '/databases/', variables('databaseName'))]"
+                            ],
+                            "properties":
+                            {
+                                "storageKey": "?",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: must be _artifactsLocationSasToken

File: 101-azure-database-migration-service/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+                        "startIpAddress": "0.0.0.0"
+                    },
+                    "dependsOn": [
+                        "[concat('Microsoft.Sql/servers/', concat(variables('targetServerName')))]"
+                    ]
+                }
+            ]
+        },
+        {
+            "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: update apiVersion

File: 101-azure-database-migration-service/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+            "type": "Microsoft.Compute/virtualMachines/extensions",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: move to a child resource

File: 101-azure-database-migration-service/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+			"defaultValue": "[resourceGroup().location]",
+			"type": "String",
+			"metadata":
+			{
+				"description": "Location where the resources will be deployed."
+			}
+		},
+        "createPublicIP":
+        {
+            "defaultValue": false,
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: use true for the sample, as people likely want to see the VM when done

File: 101-azure-database-migration-service/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+                            "enabled": true
+                        }
+                    },
+                    "keySource": "Microsoft.Storage"
+                }
+            }
+        },
+        {
+            "apiVersion": "2015-06-15",
+            "type": "Microsoft.Compute/virtualMachines/extensions",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: move to child resource

File: 101-azure-database-migration-service/azuredeploy.parameters.json [hunk 11]
```diff
@@ -0,0 +1,35 @@
+
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters":
+    {
+        "createPublicIP":
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.parameters.json]: remove this and use the defaultValue

File: 101-azure-database-migration-service/azuredeploy.json (hunk 1 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 2 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 3 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 4 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 5 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 6 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 7 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 9 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

File: 101-azure-database-migration-service/azuredeploy.parameters.json (hunk 11 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.parameters.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json [hunk 12]
```diff
@@ -1,93 +1,74 @@
...
+        "DMSServiceName": "[concat('DMS', uniqueString(resourceGroup().id))]",
+        "sourceServerName": "[take(concat('Source', uniqueString(resourceGroup().id)), 15)]",
+        "targetServerName": "[concat('targetservername', uniqueString(resourceGroup().id))]",
         "scriptLocation": "AddDatabaseToSqlServer.ps1 ",
-		"bakFileLocation": "AdventureWorks2016.bak",
-		"bacpacFileLocation": "AdventureWorks2016.bacpac",
-        "scriptFiles": "[createarray(concat('https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-azure-database-migration-service/', variables('scriptLocation')), concat('https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-azure-database-migration-service/', variables('bakFileLocation')))]",
+        "bakFileLocation": "AdventureWorks2016.bak",
+        "bacpacFileLocation": "AdventureWorks2016.bacpac",
+        "scriptFiles": "[createarray(concat(parameters('_artifactsLocation'), variables('scriptLocation')), concat(parameters('_artifactsLocation'), variables('bakFileLocation')))]",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: bak file should also be part of the sample

File: 101-azure-database-migration-service/azuredeploy.json [hunk 13]
```diff
@@ -137,81 +112,193 @@
...
+                        "[resourceId('Microsoft.DataMigration/services', variables('DMSServiceName'))]"
+                    ]
+                }
+            ],
             "dependsOn": [
                 "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('adVNet'), variables('defaultSubnetName'))]"
             ]
         },
         {
+            "condition": "[parameters('createPublicIP')]",
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: this needs to be combined into a single resource...  see: https://github.com/Azure/azure-quickstart-templates/blob/master/100-marketplace-sample/azuredeploy.json#L273 for how to handle the conditional publicIP

File: 101-azure-database-migration-service/azuredeploy.json [hunk 14]
```diff
@@ -137,81 +112,193 @@
...
                         "enableAutomaticUpdates": true
                     }
                 },
-                "networkProfile":
-                {
+                "networkProfile": {
                     "networkInterfaces": [
                         {
-                            "id": "[resourceId('Microsoft.Network/networkInterfaces', variables('sourceNicName'))]"
+                            "id": "[if(parameters('createPublicIP'), resourceId('Microsoft.Network/networkInterfaces', concat(variables('sourceNicName'), 'public')), resourceId('Microsoft.Network/networkInterfaces', concat(variables('sourceNicName'), 'private')))]"
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: see comment below on NIC resources

File: 101-azure-database-migration-service/azuredeploy.json [hunk 15]
```diff
@@ -137,81 +112,193 @@
...
+                            },
+                            "primary": true,
+                            "privateIPAddressVersion": "IPv4"
+                        }
+                    }
+                ],
+                "enableAcceleratedNetworking": false,
+                "enableIPForwarding": false
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.json]: move all of the dependsOn above the properties object in the resources

File: 101-azure-database-migration-service/azuredeploy.parameters.json [hunk 16]
```diff
@@ -1,34 +1,23 @@
...
 {
     "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
     "contentVersion": "1.0.0.0",
-    "parameters":
-    {
-        "createPublicIP":
-        {
-            "value": false
+    "parameters": {
+        "createPublicIP": {
```
Comment (bmoore-msft) [inline 101-azure-database-migration-service/azuredeploy.parameters.json]: remove this and use a defaultValue in the template

File: 101-azure-database-migration-service/azuredeploy.json [hunk 17]
```diff
@@ -1,93 +1,74 @@
...
-			    "description": "The sasToken required to access _artifactsLocation.  When the template is deployed using the accompanying scripts, a sasToken will be automatically generated. Use the defaultValue if the staging location is not secured."
-		    },
-		    "defaultValue": ""
-	    }
+        "_artifactsLocation": {
+            "type": "string",
+            "metadata": {
+                "description": "The base URI where artifacts required by this template are located including a trailing '/'"
+            },
+            "defaultValue": "https://datamigrationwebpi.blob.core.windows.net/templatefiles/"
```
[the remark this answers is not in the document]
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: The thing is I can use the GitHub repo link for the bak and script file but the bacpac file has to come from a storage account and it need a storage key.  So I am using a public container "https://datamigrationwebpi.blob.core.windows.net/templatefiles/" to save all resources

File: 101-azure-database-migration-service/azuredeploy.json (hunk 13 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: I tried using "id": "[if(parameters('createPublicIP'), resourceId('Microsoft.Network/publicIPAddresses/', variables('publicIPSourceServer')), json('null'))]"  but I get an error saying

{
  "error": {
    "code": "InvalidRequestFormat",
    "message": "Cannot parse the request.",
    "details": [
      {
        "code": "MissingJsonReferenceId",
        "message": "Value for reference id is missing. Path properties.ipConfigurations[0].properties.publicIPAddress."
      }
    ]
  }
}

when I select false for createpublicIp paramater

File: 101-azure-database-migration-service/azuredeploy.parameters.json (hunk 16 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.parameters.json]: Done

File: 101-azure-database-migration-service/azuredeploy.json (hunk 15 above)
Comment (ash1625) [reply 101-azure-database-migration-service/azuredeploy.json]: Done

**Pull request merged**